pkg/storage/stores/shipper/bloomshipper/block_downloader.go

```diff
@@ -0,0 +1,224 @@
+package bloomshipper
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	v1 "github.com/grafana/loki/pkg/storage/bloom/v1"
+	"github.com/grafana/loki/pkg/util"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/grafana/loki/pkg/queue"
+	"github.com/grafana/loki/pkg/storage/stores/shipper/bloomshipper/config"
+	"github.com/grafana/loki/pkg/util/constants"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type blockDownloader struct {
+	logger log.Logger
+
+	workingDirectory   string
+	queueMetrics       *queue.Metrics
+	queue              *queue.RequestQueue
+	blockClient        BlockClient
+	limits             Limits
+	activeUsersService *util.ActiveUsersCleanupService
+}
+
+func newBlockDownloader(config config.Config, blockClient BlockClient, limits Limits, logger log.Logger, reg prometheus.Registerer) *blockDownloader {
+	queueMetrics := queue.NewMetrics(reg, constants.Loki, "bloom_blocks_downloader")
+	//add cleanup service
+	downloadingQueue := queue.NewRequestQueue(config.BlocksDownloadingQueue.MaxTasksEnqueuedPerTenant, time.Minute, queueMetrics)
+	activeUsersService := util.NewActiveUsersCleanupWithDefaultValues(queueMetrics.Cleanup)
+	err := activeUsersService.StartAsync(context.Background())
+	if err != nil {
+		return nil
+	}
+	// If cleaner stops or fail, we will simply not clean the metrics for inactive users.
+	b := &blockDownloader{
+		logger:             logger,
+		workingDirectory:   config.WorkingDirectory,
+		queueMetrics:       queueMetrics,
+		queue:              downloadingQueue,
+		blockClient:        blockClient,
+		activeUsersService: activeUsersService,
+		limits:             limits,
+	}
+
+	for i := 0; i < config.BlocksDownloadingQueue.WorkersCount; i++ {
+		go b.serveDownloadingTasks(fmt.Sprintf("worker-%d", i))
```

Comment:
agree, but be careful with the `for ctx.Err` bit as explained in detail [here](https://github.com/grafana/loki/pull/11181#discussion_r1393498266). It should use `select` to avoid saturating the cpu.

Reply:
I expected queue to be stopped in this case and the worker would receive `ErrStopped` and would close the goroutine 
https://github.com/grafana/loki/blob/27411cff087e6fa2c660c6bfa577c31945ab5232/pkg/queue/queue.go#L146 
wdyt?

Reply:
ok, but it looks like we do not call StopAsync on the queue ;)

Reply:
I managed to achieve the workers shutdown without `select` statement, because `dequeue`  function also checks if the context is not canceled.  
https://github.com/grafana/loki/pull/11201/commits/6b05ec5e236b8747f9c0222bd318f7963419fe7b